cpp/src/wholememory/memory_handle.cpp

```diff
@@ -1120,6 +1136,168 @@ class chunked_device_wholememory_impl : public wholememory_impl {
   wholememory_gref_t gref_;
 };
 
+#ifdef WITH_NVSHMEM_SUPPORT
+class nvshmem_device_wholememory_impl : public wholememory_impl {
+ public:
+  nvshmem_device_wholememory_impl(wholememory_handle_t wholememory_handle,
+                                  size_t total_size,
+                                  wholememory_comm_t comm,
+                                  wholememory_memory_type_t memory_type,
+                                  wholememory_memory_location_t memory_location,
+                                  size_t data_granularity)
+    : wholememory_impl(
+        wholememory_handle, total_size, comm, memory_type, memory_location, data_granularity)
+  {
+    WHOLEMEMORY_CHECK(type_ == WHOLEMEMORY_MT_DISTRIBUTED);
+    WHOLEMEMORY_CHECK(location_ == WHOLEMEMORY_ML_DEVICE);
+    WHOLEMEMORY_CHECK(comm->preferred_distributed_backend == WHOLEMEMORY_DB_NVSHMEM);
+    distrubuted_backend_ = WHOLEMEMORY_DB_NVSHMEM;
+    check_or_set_nvshmem_heap_kind();
+    init_nvshmem_with_comm(comm);
+    WHOLEMEMORY_CHECK(comm->bind_to_nvshmem);
+  }
+
+  void create_memory() override
+  {
+    each_rank_same_chunk_strategy();
+    generate_rank_partition_strategy();
+    nvshmem_malloc_device_memory();
+    register_nvshmem_device_memory();
+  }
+
+  void destroy_memory() noexcept override
+  {
+    unregister_nvshmem_device_memory();
+    nvshmem_free_device_memory();
+  }
+  // [[nodiscard]] wholememory_gref_t get_global_reference() const noexcept override { return gref_;
+  // }
+
+  bool contains_pointer(const void* ptr) const override
+  {
+    uint64_t int_ptr = reinterpret_cast<uint64_t>(ptr);
+    size_t acc_size  = 0;
+
+    // we should only handle address that can be accessed by p2p.
+    // TODO: here return true ? // or print warning info ...
+    for (int i = 0; i < comm_->world_size; i++) {
+      size_t mem_size_of_this_rank_and_after = total_size_ - acc_size;
+      size_t mem_size_for_current_rank =
+        std::min(mem_size_of_this_rank_and_after, rank_partition_strategy_.partition_mem_stride);
+      acc_size += mem_size_for_current_rank;
+      uint64_t int_start_ptr =
+        reinterpret_cast<uint64_t>(nvshmem_ptr(nvshmem_memory_handle_.local_alloc_mem_ptr, i));
+      if (int_start_ptr == 0) continue;
+      if (int_ptr >= int_start_ptr && int_ptr < int_start_ptr + mem_size_for_current_rank) {
+        return true;
+      }
+    }
```

Comment:
If you have intention to modernize the codebase overtime, you may consider using `std::any_of`. This better documents the intention of this code (if any_of `int_ptr` satisfies the condition, return true, otherwise, return false).